src/modalities/models/huggingface/collator.py

```diff
@@ -0,0 +1,226 @@
+        labels_mask = ~mask_indices
+
+        input_ids_sentinel = self.create_sentinel_ids(mask_indices.astype(np.int8))
+        labels_sentinel = self.create_sentinel_ids(labels_mask.astype(np.int8))
+
+        input_ids = self.filter_input_ids(sample_tensor, input_ids_sentinel)
+        label_ids = self.filter_input_ids(sample_tensor, labels_sentinel)
+
+        samples = {self.sample_key: torch.from_numpy(input_ids)}
+        targets = {self.target_key: torch.from_numpy(label_ids)}
+
+        return DatasetBatch(targets=targets, samples=samples)
+
+    def create_sentinel_ids(self, mask_indices: np.ndarray[bool]) -> np.ndarray[int]:
+        """
+        Sentinel ids creation given the indices that should be masked.
+        The start indices of each mask are replaced by the sentinel ids in increasing
+        order. Consecutive mask indices to be deleted are replaced with `-1`.
+        Sentinel ids are defined by the tokenizer as 32099: <extra_id_0>, 32098: <extra_id_1>, ..., 32000: <extra_id_99>
+
+        Args:
+            mask_indices (np.ndarray[bool]): binary mask of tokens to mask out
+
+        Returns:
+            np.ndarray[int]: array of sentinel ids at starting point of spans to be masked out
+        """
+        start_indices = mask_indices - np.roll(mask_indices, 1, axis=-1) * mask_indices
+        start_indices[:, 0] = mask_indices[:, 0]
+
+        sentinel_ids = np.where(start_indices != 0, np.cumsum(start_indices, axis=-1), start_indices)
+        sentinel_ids = np.where(sentinel_ids != 0, (self.tokenizer.vocab_size - sentinel_ids), 0)
+        sentinel_ids -= mask_indices - start_indices
+
+        return sentinel_ids
+
+    def filter_input_ids(self, input_ids: np.ndarray[int], sentinel_ids: np.ndarray[int]) -> np.ndarray[int]:
+        """
+        Puts sentinel mask on `input_ids` and fuse consecutive mask tokens into a single mask token by deleting.
+        This will reduce the sequence length from `expanded_inputs_length` to `input_length`.
+        Sequence length after masking is deterministic: all masked sequences will have the same length.
+
+        Args:
+            input_ids (np.ndarray[int]): sequence of input token ids
+            sentinel_ids (np.ndarray[int]): array of sentinel ids at starting point of spans to be masked out
+
+        Returns:
+            np.ndarray[int]: input ids with masked out spans replaced by single sentinel token
+        """
+        batch_size = input_ids.shape[0]
+
+        input_ids_full = np.where(sentinel_ids != 0, sentinel_ids, input_ids)
+        # input_ids tokens and sentinel tokens are >= 0, tokens < 0 are
+        # masked tokens coming after sentinel tokens and should be removed
+        input_ids = input_ids_full[input_ids_full >= 0].reshape((batch_size, -1))
+        input_ids = np.concatenate(
+            [input_ids, np.full((batch_size, 1), self.tokenizer.tokenizer.eos_token_id)], axis=-1, dtype=np.int32
```

Comment:
Need to verify that we actually need the `eos_token_id` here. I would assume that the pbin file already comes with the eos_token for every document.

Reply:
Turns out this adds an EOS token at the end of every batch, not document, which seems clearly bad. So removed it